styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -277,15 +259,64 @@ private void dequeueInstance(Map<String, Resource> resources,
           Event.runError(instanceState.workflowInstance(),
               String.format("Referenced resources not found: %s", unknownResources)),
           instanceState.runState().counter());
-    } else {
-      double sleepingTime = dequeueRateLimiter.acquire();
-      if (sleepingTime > 0.0001) {
-        LOG.debug("Dequeue rate limited and slept for {} ms", sleepingTime * 1000);
+      return;
+    }
+
+    // Check resource limits. This is racy and can give false positives but the transactional
+    // checking happens later. This is just intended to avoid spinning on exhausted resources.
+    final List<String> exhaustedResources = instanceResourceRefs.stream()
+        .filter(this::limitReached)
+        .sorted()
+        .collect(toList());
+    if (!exhaustedResources.isEmpty()) {
+      LOG.debug("Resource limit reached for instance, not dequeueing: {}: exhausted resources={}",
+          instanceState.workflowInstance(), exhaustedResources );
+      return;
+    }
+
+    // Check for execution blocker
+    if (config.executionGatingEnabled()) {
+      Optional<ExecutionBlocker> blocker = Optional.empty();
+      try {
+        blocker = gate.executionBlocker(instanceState.workflowInstance())
+            .toCompletableFuture().get(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException(e);
+      } catch (ExecutionException | TimeoutException e) {
+        LOG.warn("Failed to check execution blocker for {}, assuming there is no blocker",
+            instanceState.workflowInstance(), e);
+      }
+
+      if (blocker.isPresent()) {
+        stateManager.receiveIgnoreClosed(Event.retryAfter(instanceState.workflowInstance(),
+            blocker.get().delay().toMillis()),
+            instanceState.runState().counter());
+        LOG.debug("Dequeue rescheduled: {}: {}", instanceState.workflowInstance(), blocker.get());
+        return;
       }
+    }
 
-      // Racy: some resources may have been removed (become unknown) by now; in that case the
-      // counters code during dequeue will treat them as unlimited...
-      sendDequeue(instanceState, instanceResourceRefs);
+    double sleepingTime = dequeueRateLimiter.acquire();
+    if (sleepingTime > 0.0001) {
+      LOG.debug("Dequeue rate limited and slept for {} ms", sleepingTime * 1000);
+    }
+
+    // Racy: some resources may have been removed (become unknown) by now; in that case the
+    // counters code during dequeue will treat them as unlimited...
+    sendDequeue(instanceState, instanceResourceRefs);
+  }
+
+  private boolean limitReached(final String resourceId) {
+    try {
+      final CounterSnapshot counterSnapshot = shardedCounter.getCounterSnapshot(resourceId);
```

Comment:
The CounterSnapshot was meant to be private as an implementation detail, made public to ease testing at some point, and so it goes :)

Reply:
I guess we could expose a nicer api for doing this check instead of interacting directly with the `CounterSnapshot`.